File: packages/agents-python/berth_agents/providers/azure_openai.py
"""Azure OpenAI isn't just OpenAI with a different base_url — it
authenticates with an `api-key` header (or Entra ID) instead of `Authorization:
Bearer`, routes by a deployment name in the URL path rather than `model` in
the request body, and requires an `api-version` query param on every
request. `openai`'s own `AsyncAzureOpenAI` client already handles all three;
this just reuses the same chat()/chat_stream() implementation every
OpenAI-shaped provider shares. Mirrors @berth/agents' providers/azure-openai.ts."""

from __future__ import annotations

import os

from openai import AsyncAzureOpenAI

from .openai import _OpenAIProvider

DEFAULT_API_VERSION = "2024-10-21"


def create_azure_openai_provider(
    *,
    api_key: str | None = None,
    endpoint: str | None = None,
    deployment: str | None = None,
    api_version: str | None = None,
) -> _OpenAIProvider:
    resolved_deployment = deployment or os.environ.get("AZURE_OPENAI_DEPLOYMENT")
    if not resolved_deployment:
        raise ValueError("create_azure_openai_provider() needs deployment= (or AZURE_OPENAI_DEPLOYMENT) — Azure routes by deployment name, not model")

    client = AsyncAzureOpenAI(
        api_key=api_key or os.environ.get("AZURE_OPENAI_API_KEY"),
        azure_endpoint=endpoint or os.environ.get("AZURE_OPENAI_ENDPOINT"),
        azure_deployment=resolved_deployment,
        api_version=api_version or os.environ.get("AZURE_OPENAI_API_VERSION") or DEFAULT_API_VERSION,
    )
    # Azure's chat.completions.create() ignores `model` in favor of the
    # client's configured deployment, but the field is still required by the
    # SDK's own types — the deployment name is a reasonable value to send.
    return _OpenAIProvider(client, resolved_deployment, name="azure-openai")

File: packages/agents/src/providers/azure-openai.ts
import { AzureOpenAI } from "openai";
import type { LLMProvider } from "../types.js";
import { createOpenAICompatibleProvider } from "./openai.js";

export interface AzureOpenAIProviderOptions {
  apiKey?: string;
  /** Your Azure OpenAI resource endpoint, e.g. `https://<resource>.openai.azure.com`. Defaults to `AZURE_OPENAI_ENDPOINT`. */
  endpoint?: string;
  /** The deployment name you created in Azure (not the underlying model's own name) — Azure routes by deployment, not by `model` in the request body. Defaults to `AZURE_OPENAI_DEPLOYMENT`. */
  deployment?: string;
  /** Azure's API version query param. Defaults to `AZURE_OPENAI_API_VERSION`, or a recent stable version if neither is set. */
  apiVersion?: string;
}

const DEFAULT_API_VERSION = "2024-10-21";

/**
 * Azure OpenAI isn't just OpenAI with a different `baseURL` — it authenticates
 * with an `api-key` header (or Entra ID) instead of `Authorization: Bearer`,
 * routes by a deployment name in the URL path rather than `model` in the
 * request body, and requires an `api-version` query param on every request.
 * The `openai` package's own `AzureOpenAI` client (re-exported from its main
 * entrypoint) already handles all three; this just wires it into the same
 * chat()/chatStream() implementation every other OpenAI-shaped provider
 * shares (see `createOpenAICompatibleProvider()` in `openai.ts`).
 */
export function createAzureOpenAIProvider(options: AzureOpenAIProviderOptions = {}): LLMProvider {
  const deployment = options.deployment ?? process.env.AZURE_OPENAI_DEPLOYMENT;
  if (!deployment) {
    throw new Error("createAzureOpenAIProvider() needs `deployment` (or AZURE_OPENAI_DEPLOYMENT) — Azure routes by deployment name, not model");
  }
  const client = new AzureOpenAI({
    apiKey: options.apiKey ?? process.env.AZURE_OPENAI_API_KEY,
    endpoint: options.endpoint ?? process.env.AZURE_OPENAI_ENDPOINT,
    deployment,
    apiVersion: options.apiVersion ?? process.env.AZURE_OPENAI_API_VERSION ?? DEFAULT_API_VERSION,
  });
  // Azure's chat.completions.create() ignores `model` in favor of the
  // client's configured deployment, but the field is still required by the
  // SDK's own types — the deployment name is a reasonable value to send.
  return createOpenAICompatibleProvider(client, deployment, "azure-openai");
}
